subs/pantry/src/Pantry/HPack.hs

```diff
@@ -0,0 +1,92 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Pantry.HPack (findOrGenerateCabalFile, hpack) where
+
+import RIO
+import RIO.Process
+import Pantry.Types
+import qualified RIO.List as List
+import qualified RIO.FilePath as FilePath
+import qualified Hpack
+import qualified Hpack.Config as Hpack
+import Path (Path, Abs, File, toFilePath, Dir, (</>), filename, parseAbsDir, parent, parseRelFile, fromAbsFile)
+import Path.IO (doesFileExist, resolveDir', listDir)
+
+-- | Get the filename for the cabal file in the given directory.
+--
+-- If no .cabal file is present, or more than one is present, an exception is
+-- thrown via 'throwM'.
+--
+-- If the directory contains a file named package.yaml, hpack is used to
+-- generate a .cabal file from it.
+findOrGenerateCabalFile
+    :: forall env. (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+    => Path Abs Dir -- ^ package directory
+    -> RIO env (PackageName, Path Abs File)
+findOrGenerateCabalFile pkgDir = do
+    hpack pkgDir
+    files <- filter (flip hasExtension "cabal" . toFilePath) . snd
+         <$> listDir pkgDir
+    -- If there are multiple files, ignore files that start with
+    -- ".". On unixlike environments these are hidden, and this
+    -- character is not valid in package names. The main goal is
+    -- to ignore emacs lock files - see
+    -- https://github.com/commercialhaskell/stack/issues/1897.
+    let isHidden ('.':_) = True
+        isHidden _ = False
+    case filter (not . isHidden . toFilePath . filename) files of
+        [] -> throwIO $ NoCabalFileFound pkgDir
+        [x] -> maybe
+          (throwIO $ InvalidCabalFilePath x)
+          (\pn -> pure $ (pn, x)) $
+            List.stripSuffix ".cabal" (toFilePath (filename x)) >>=
+            parsePackageName
+        _:_ -> throwIO $ MultipleCabalFilesFound pkgDir files
+      where hasExtension fp x = FilePath.takeExtension fp == "." ++ x
+
+-- getCabalBS :: ByteString -> IO ByteString
+-- getCabalBS hpackBS = do
+--   withSystemTempDirectory "temp-hpack-dir" $
+--                               \tmpdir -> withWorkingDir tmpdir $ do
+--                                            B.writeFile Hpack.packageConfig hpackBS
+--                                            hpack undefined
+
+
+-- | Generate .cabal file from package.yaml, if necessary.
+hpack
+  :: (HasPantryConfig env, HasLogFunc env, HasProcessContext env)
+  => Path Abs Dir
+  -> RIO env ()
+hpack pkgDir = do
+    packageConfigRelFile <- parseRelFile Hpack.packageConfig
+    let hpackFile = pkgDir Path.</> packageConfigRelFile
+    exists <- liftIO $ doesFileExist hpackFile
+    when exists $ do
```

Comment:
There is a cool thing in RIO: `whenM` and `doesFileExist` imported from `RIO.Directory` is restricted to `MonadIO`, so you can rewrite it as:

```haskell
whenM (doesFileExist hpackFile) $ do
   ...
```